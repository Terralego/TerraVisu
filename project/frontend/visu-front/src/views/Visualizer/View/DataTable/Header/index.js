import Header from './Header';
import { connectView } from '../../context';

export default connectView(({ setLayerState, layersTreeState }) => ({
  setLayerState,
  displayedLayer: Array
    .from(layersTreeState)
    .find(([, { table }]) => table),
}))(Header);
