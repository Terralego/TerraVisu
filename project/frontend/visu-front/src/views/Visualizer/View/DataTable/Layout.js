import React from 'react';
import classnames from 'classnames';

export const Layout = ({ visible, children }) => (
  <div className={classnames({
    'data-table': true,
    'data-table--visible': visible,
  })}
  >
    {visible && children}
  </div>
);

export default Layout;
