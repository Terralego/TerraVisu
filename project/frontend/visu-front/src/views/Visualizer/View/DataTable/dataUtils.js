/**
 * Round value with decimals
 *
 * @param {number} value
 * @param {number} decimals
 * @returns {number}
 */
const round = (value, decimals) => (Math.round(value * (10 ** decimals)) / (10 ** decimals));

export const extractColumns = (fields = [], results) => {
  if (fields.length) {
    return fields.map(({ value, label, display = true, sortable = true, ...props }) => ({
      value,
      label,
      display,
      sortable,
      ...props,
    }));
  }

  if (!results[0]) return [];

  const [{ _source: source }] = results;

  return Object.keys(source).map(column => ({
    value: column,
    display: true,
    sortable: true,
  }));
};

/**
 * Format value according to settings
 *
 * @param {string|number} value
 * @param {{}} settings Object of form {decimals: 1}
 * @returns {number|string}
 */
export const formatFromSettings = (value, settings) => {
  const { decimals } = settings;
  const isNumber = typeof value === 'number';
  let formattedData = value;

  if (isNumber) {
    if (decimals) {
      formattedData = round(formattedData, decimals);
    }
  }
  return formattedData;
};

/**
 * Get datum from ElasticSearch values and format it according to settings
 *
 * @param {{}} dataSource Values object fetch from ElasticSearch
 * @param {string} key Datum key to get
 * @param {{}} settings Object of form {decimals: 1}
 * @returns {string|string|*}
 */
export const getData = (dataSource, key, settings) => {
  let value = dataSource[key];
  const { objectIndent = 2 } = settings;
  if (Array.isArray(value)) return value.join(',');
  if (value && typeof value === 'object') return JSON.stringify(value, null, objectIndent);
  value = [null, undefined].includes(value) ? '' : value;
  return settings ? formatFromSettings(value, settings) : value;
};

/**
 * Map data to columns for display while interpolating a value from source if needed.
 *
 * @param {{}[]} fields List of columns objects {value: property key}
 * @param {{}[]} results List of objects containing id and source
 * @param {{}} settings Object of form {decimals: 1}
 * @returns {{}[][]} Each result mapped to each column
 */
export const prepareData = (fields = [], results, settings) => {
  const properties = fields.map(({ value }) => value);

  return results
    .map(({ _id, _source: dataSource }) => properties.map(key => {
      const interpolation = key.match(/{([^}]+)}/g);
      if (interpolation) {
        return key.replace(/{([^}]+)}/g, (match, p1) => getData(dataSource, p1, settings));
      }
      return getData({ _id, ...dataSource }, key, settings);
    }));
};

/**
 * Export data in xls from state
 *
 * @param {string} name Filename
 * @param {[]} data
 * @param {function} callback Function to be called just before saving sheet
 * @returns {Promise<void>}
 */
export const exportSpreadsheet = async ({ name, data, callback }) => {
  const xlsx = await import('xlsx');
  const workbook = xlsx.utils.book_new();
  const sheet = xlsx.utils.aoa_to_sheet(data);

  if (callback) {
    callback(xlsx, sheet, workbook);
  }

  // xslx has a hard limit to 31 non-special chars
  // see https://support.office.com/en-us/article/Rename-a-worksheet-3F1F7148-EE83-404D-8EF0-9FF99FBAD1F9
  const cleanedName = name.replace(/[\][*?/\\:]/gi, '').substring(0, 30);
  xlsx.utils.book_append_sheet(workbook, sheet, cleanedName);
  xlsx.writeFile(workbook, `${name}.xlsx`);
};

export default { extractColumns, prepareData };
