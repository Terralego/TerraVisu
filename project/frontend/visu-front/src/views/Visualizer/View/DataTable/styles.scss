@import '../../../variables.scss';

.data-table {
  flex-shrink: 0;
  padding: 0;
  height: 0;
  transition: height .2s ease-in;
  display: flex;
  position: relative;
  height: 0;
  z-index: 80;
  will-change: height;

  &--visible {
    height: 33vh;
  }

  .component-table {
    outline: none;
  }

  .table-container {
    transition: transform .2s ease-in;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    margin: 0;
    transform: translate3d(0, 100%, 0);

    &--visible {
      transform: translate3d(0, 0, 0);
    }

    &--full.table-container--visible {
      height: calc(100vh);
      transform: translate3d(0, -67.25vh, 0);
    }

    &--is-resizing {
      height: calc(100vh);
    }
  }

  .table {
    display: flex;
    flex-direction: column;
  }

  .table-header {
    background: #F5F8FA;
    padding: .8rem;
    position: relative;
    display: flex;
    align-items: center;

    & > * {
      width: auto;
    }

    &__title {
      text-align: center;
      width: 100%;
      display: flex;
      justify-content: center;
    }

    &__counter {
      position: relative;
      margin-left: 1rem;
    }

    &__actions {
      position: absolute;
      right: .6rem;
    }

    &__loading {
      position: absolute;
      left: 50%;
      transform: translateX(-50%);
    }

    &--is-loading {
      .table-header__count {
        visibility: hidden;
      }
    }
  }
}

.table-columns-selector__toggle-all {
  font-style: italic;
}

@media(min-width: $breakpoints-tablet) and (max-width: $breakpoints-desktop) and (orientation: landscape) {
  .data-table .table-header {
    padding: .8rem .5rem;

    &__title {
      min-width: 0;
      max-width: 65%;

      &--label {
        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
        margin: 0;
      }
    }

    &__counter {
      margin-left: .5rem;
    }
  }
}

@media(min-width: $breakpoints-phone) and (max-width: $breakpoints-tablet) and (orientation: portrait) {
  .data-table {
    display: none;
  }
}
