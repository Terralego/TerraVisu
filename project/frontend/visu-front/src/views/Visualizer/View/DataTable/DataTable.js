import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import Table from '@terralego/core/modules/Table';
import searchService, {
  getExtent,
  getSearchParamFromProperty,
} from '@terralego/core/modules/Visualizer/services/search';
import debounce from 'debounce';

import { extractColumns, prepareData, exportSpreadsheet } from './dataUtils';
import Header from './Header';

import './styles.scss';

const sort = (data, column = 0) => data
  .map((line, index) => ({
    value: line[column], index,
  }))
  .sort((a, b) => a.value - b.value)
  .map(({ index }) => data[index]);

export class DataTable extends React.Component {
  static propTypes = {
    displayedLayer: PropTypes.shape({
      filters: PropTypes.shape({
        layer: PropTypes.string.isRequired,
        exportable: PropTypes.bool,
        table: PropTypes.shape({
          title: PropTypes.string,
        }),
        fields: PropTypes.arrayOf(PropTypes.shape({
          value: PropTypes.string.isRequired,
          label: PropTypes.string.isRequired,
          exportable: PropTypes.bool,
          display: PropTypes.bool, // default true
        })),
      }),
    }),
    isTableVisible: PropTypes.bool,
    query: PropTypes.string,
    filters: PropTypes.shape({
      // anyKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    }),
  };

  static defaultProps = {
    displayedLayer: undefined,
    isTableVisible: true,
    query: '',
    filters: {},
  };

  state = {
    columns: null,
    results: [],
    resultsTotal: 0,
    loading: true,
    selectedFeatures: [],
  };

  debouncedLoadResults = debounce(() => this.loadResults(), 500);

  componentDidMount () {
    const { displayedLayer } = this.props;
    if (displayedLayer) {
      this.debouncedLoadResults();
    }
  }

  shouldComponentUpdate ({ isTableVisible }) {
    const {
      isTableVisible: prevTableVisible,
    } = this.props;

    if (isTableVisible !== prevTableVisible || isTableVisible) {
      return true;
    }
    return false;
  }

  componentDidUpdate ({
    displayedLayer: {
      filters: { layer: prevLayer, fields: prevFields } = {},
      state: { filters: prevFilters } = {},
    } = {},
    query: prevQuery,
  }, {
    extent: prevExtent,
  }) {
    const {
      displayedLayer,
      displayedLayer: {
        filters: { layer, fields } = {},
        state: { filters } = {},
      } = {},
      query,
    } = this.props;

    const { extent } = this.state;

    if (displayedLayer) {
      if (layer !== prevLayer
       || prevFields !== fields) {
        this.resetColumns();
      }

      if (layer !== prevLayer
        || query !== prevQuery
        || filters !== prevFilters
        || (extent && this.extentChanged())
        || prevFields !== fields
        || extent !== prevExtent) { // This test must keep at last position
        this.debouncedLoadResults();
      }
    }
  }

  resize = () => this.setState(({ full: prevFull }) => {
    const full = !prevFull;

    if (full) {
      setTimeout(() => this.setState({ full, isResizing: false }), 300);
      return { isResizing: true };
    }
    setTimeout(() => this.setState({ isResizing: false }), 300);
    return { full, isResizing: true };
  })

  toggleExtent = () => this.setState(({ extent }) => ({ extent: !extent }));

  onSelection = selection => {
    const { features } = this.state;

    if (!features) return;
    const { results } = this.state;
    const selectedFeatures = selection.map(id => ({ _id: results[id][0] }));
    this.setState({ selectedFeatures });
  }

  exportDataAction = () => {
    const {
      displayedLayer: {
        label: name,
        filters: { fields = [] } = {},
      },
      exportCallback,
    } = this.props;

    if (!fields) return;

    const { columns, results } = this.state;

    // Find out which column (indexes) are exportable
    const exportableColumnIndexes = columns.reduce((store, { value }, index) => {
      const fieldConfig = fields.find(field => field.value === value);
      const { exportable = false } = fieldConfig || {};
      return exportable ? [...store, index] : store;
    }, []);

    // Create array of column header texts
    const columnLabels = columns.map(({ value, label = value }) => label);

    // Go through each line and keep only exportable columns
    const data = [columnLabels, ...results]
      .map(dataLine => exportableColumnIndexes.map(index => dataLine[index]));

    exportSpreadsheet({
      name,
      data,
      callback: exportCallback,
    });
  };

  resetColumns () {
    this.setState({ columns: null });
  }

  extentChanged () {
    const { map, visibleBoundingBox } = this.props;
    const prevExtent = this.prevExtent || [[], []];
    this.prevExtent = getExtent(map, visibleBoundingBox);
    const [[a, b], [c, d]] = prevExtent;
    const [[aa, bb], [cc, dd]] = this.prevExtent;
    return !(a === aa && b === bb && c === cc && d === dd);
  }

  async loadResults () {
    const {
      displayedLayer: {
        filters: { layer, fields, form } = {},
        state: { filters = {} } = {},
        baseEsQuery,
      } = {},
      query,
      map,
      visibleBoundingBox,
    } = this.props;
    const { columns, extent } = this.state;

    this.setState({ loading: true });

    this.prevExtent = extent && getExtent(map, visibleBoundingBox);
    const boundingBox = this.prevExtent;

    const properties = {
      ...Object.keys(filters).reduce((all, key) => ({
        ...all,
        ...getSearchParamFromProperty(filters, form, key),
      }), {}),
    };

    const resp = await searchService.search({
      index: layer,
      query,
      properties,
      boundingBox,
      baseQuery: baseEsQuery,
      include: fields && fields.reduce((all, { value }) => {
        const interpolation = value.match(/\{[^}]+\}/g);
        return [
          ...all,
          ...interpolation
            ? interpolation.map(match => match.match(/\{([^}]+)\}/)[1])
            : [value],
        ];
      }, []),
    });

    const { hits: { hits, total: { value: resultsTotal } } } = resp;
    const newColumns = columns || extractColumns(fields, hits);

    const results = sort(prepareData(newColumns, hits, {
      decimals: 1,
    }));

    this.setState({ features: hits, results, resultsTotal, columns: newColumns, loading: false });
  }

  render () {
    const { displayedLayer, isTableVisible } = this.props;

    const {
      label,
      compare,
      filters: {
        layer,
        table: { title } = {},
        exportable,
        fields = [],
      } = {},
    } = displayedLayer || {};

    const haveExportableField = fields.some(({ exportable: exportableField }) => exportableField);
    const showExportButton = exportable && haveExportableField;

    const {
      columns, results, resultsTotal,
      loading, full, isResizing,
      extent, selectedFeatures,
    } = this.state;
    const firstLoading = !columns && loading;

    const { resize, toggleExtent, onSelection, exportDataAction } = this;
    return (
      <div className={classnames({
        'data-table': true,
        'data-table--visible': isTableVisible,
      })}
      >
        <div
          className={classnames({
            'table-container': true,
            'table-container--visible': isTableVisible,
            'table-container--is-resizing': isResizing,
            'table-container--full': full,
          })}
        >
          {isTableVisible && (
            <Table
              columns={columns || []}
              data={results || []}
              onSelection={onSelection}
              Header={props => (
                <Header
                  {...props}
                  resultsTotal={resultsTotal}
                  toggleExtent={toggleExtent}
                  extent={extent}
                  layer={layer}
                  compare={compare}
                  selectedFeatures={selectedFeatures || []}
                  loading={loading}
                  title={title || label}
                  full={full}
                  resize={resize}
                  exportData={showExportButton && exportDataAction}
                />
              )}
              locales={{
                sortAsc: 'Trier dans l\'ordre croissant',
                sortDesc: 'Trier dans l\'ordre dÃ©croissant',
              }}
              loading={firstLoading}
            />
          )}
        </div>
      </div>
    );
  }
}

export default DataTable;
