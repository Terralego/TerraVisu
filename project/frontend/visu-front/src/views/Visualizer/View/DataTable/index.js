import { connectView } from '../context';

import DataTable from './DataTable';

export default connectView(({ layersTreeState, query, map, visibleBoundingBox }) => ({
  query,
  map,
  visibleBoundingBox,
  displayedLayer: Array
    .from(layersTreeState)
    .filter(([, { table }]) => table)
    .map(([layer, state]) => ({ ...layer, state }))[0],
}))(DataTable);
