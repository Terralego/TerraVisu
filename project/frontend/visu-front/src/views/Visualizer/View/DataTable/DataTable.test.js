import React from 'react';
import renderer from 'react-test-renderer';
import DataTable from './DataTable';

jest.mock('@terralego/core/modules/Visualizer/services/search', () => ({
  search: jest.fn(() => ({
    hits: {
      hits: [{
        _source: {
          foo_prop: 'foo',
          bar_prop: 'bar',
        },
      }, {
        _source: {
          foo_prop: 'foo1',
          bar_prop: 'bar1',
        },
      }],
      total: { value: 2 },
    },
  })),
}));

jest.mock('debounce', () => fn => () => fn());

jest.mock('@terralego/core/modules/Table', () => props => <p {...props}>Table</p>);
jest.mock('@terralego/core/modules/Table/components/ColumnsSelector', () =>
  props => (
    <p {...props}>ColumnsSelector</p>
  ));

it('should render correctly', () => {
  const tree = renderer.create((
    <DataTable />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render unvisible correctly', () => {
  const tree = renderer.create((
    <DataTable isTableVisible={false} />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render correctly with results', async done => {
  const tree = renderer.create((
    <DataTable
      displayedLayer={{
        filters: {
          layer: 'foo',
          fields: [{
            label: 'Foo',
            value: 'foo_prop',
          }, {
            label: 'Bar',
            value: 'bar_prop',
          }],
        },
      }}
      resultsTotal={1}
    />
  ));

  await true;
  await true;
  await true;

  expect(tree.toJSON()).toMatchSnapshot();

  done();
});

it('should render a loader', () => {
  const tree = renderer.create((
    <DataTable
      displayedLayer={{
        filters: {
          layer: 'foo',
          fields: [{
            label: 'Foo',
            value: 'foo_prop',
          }, {
            label: 'Bar',
            value: 'bar_prop',
          }],
        },
      }}
      resultsTotal={1}
    />
  ));

  expect(tree.toJSON()).toMatchSnapshot();
});
