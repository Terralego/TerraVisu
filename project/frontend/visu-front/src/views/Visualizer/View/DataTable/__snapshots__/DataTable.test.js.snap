// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a loader 1`] = `
<div
  className="data-table data-table--visible"
>
  <div
    className="table-container table-container--visible"
  >
    <p
      Header={[Function]}
      columns={Array []}
      data={Array []}
      loading={true}
      locales={
        Object {
          "sortAsc": "Trier dans l'ordre croissant",
          "sortDesc": "Trier dans l'ordre décroissant",
        }
      }
      onSelection={[Function]}
    >
      Table
    </p>
  </div>
</div>
`;

exports[`should render correctly 1`] = `
<div
  className="data-table data-table--visible"
>
  <div
    className="table-container table-container--visible"
  >
    <p
      Header={[Function]}
      columns={Array []}
      data={Array []}
      loading={true}
      locales={
        Object {
          "sortAsc": "Trier dans l'ordre croissant",
          "sortDesc": "Trier dans l'ordre décroissant",
        }
      }
      onSelection={[Function]}
    >
      Table
    </p>
  </div>
</div>
`;

exports[`should render correctly with results 1`] = `
<div
  className="data-table data-table--visible"
>
  <div
    className="table-container table-container--visible"
  >
    <p
      Header={[Function]}
      columns={
        Array [
          Object {
            "display": true,
            "label": "Foo",
            "sortable": true,
            "value": "foo_prop",
          },
          Object {
            "display": true,
            "label": "Bar",
            "sortable": true,
            "value": "bar_prop",
          },
        ]
      }
      data={
        Array [
          Array [
            "foo",
            "bar",
          ],
          Array [
            "foo1",
            "bar1",
          ],
        ]
      }
      loading={false}
      locales={
        Object {
          "sortAsc": "Trier dans l'ordre croissant",
          "sortDesc": "Trier dans l'ordre décroissant",
        }
      }
      onSelection={[Function]}
    >
      Table
    </p>
  </div>
</div>
`;

exports[`should render unvisible correctly 1`] = `
<div
  className="data-table"
>
  <div
    className="table-container"
  />
</div>
`;
